p2p/protocols/protocol.go

```diff
@@ -222,34 +228,111 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
 		spec:   spec,
 		encode: encode,
 		decode: decode,
+		eg:     &errgroup.Group{},
 	}
 }
 
-// Run starts the forever loop that handles incoming messages
-// called within the p2p.Protocol#Run function
-// the handler argument is a function which is called for each message received
+// Run starts the forever loop that handles incoming messages.
+// The handler argument is a function which is called for each message received
 // from the remote peer, a returned error causes the loop to exit
-// resulting in disconnection
+// resulting in disconnection of the protocol
+// there can be only one run loop per peer.
+// It returns the error that caused the loop to exit.
 func (p *Peer) Run(handler func(ctx context.Context, msg interface{}) error) error {
-	for {
-		if err := p.handleIncoming(handler); err != nil {
-			if err != io.EOF {
-				metrics.GetOrRegisterCounter("peer.handleincoming.error", nil).Inc(1)
-				log.Error("peer.handleIncoming", "err", err)
+	p.setRunning(true)
+
+	done := make(chan error, 1)
+	go func() {
+		for {
+			msg, err := p.rw.ReadMsg()
+			if err != nil {
+				if err != io.EOF {
+					metrics.GetOrRegisterCounter("peer.handleincoming.error", nil).Inc(1)
+					log.Error("peer.handleIncoming", "err", err)
+				}
+
+				select {
+				case done <- err:
+				default:
+				}
+
+				return
 			}
+
+			p.dispatchAsyncHandleMsg(&msg, handler, done)
+		}
+	}()
+
+	select {
+	case err := <-done:
+		return err
+	case err := <-p.shutdown:
+		return err
+	}
+}
+
+func (p *Peer) setRunning(isRunning bool) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	p.running = isRunning
+	return
+}
+
+func (p *Peer) dispatchAsyncHandleMsg(msg *p2p.Msg, handler func(ctx context.Context, msg interface{}) error, done chan error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	if p.running {
+		p.eg.Go(func() error {
+			err := p.handleMsg(msg, handler)
+			if err != nil {
+				p.stop(err, done)
+			}
+
 			return err
+		})
+	} else {
+		select {
+		case done <- nil:
+		default:
 		}
 	}
 }
 
-// Drop disconnects a peer.
+// Drop disconnects a peer
 // TODO: may need to implement protocol drop only? don't want to kick off the peer
 // if they are useful for other protocols
 func (p *Peer) Drop(reason string) {
-	log.Error("dropping peer with DiscSubprotocolError", "peer", p.ID(), "reason", reason)
+	log.Info("dropping peer with DiscSubprotocolError", "peer", p.ID(), "reason", reason)
 	p.Disconnect(p2p.DiscSubprotocolError)
 }
 
+//Shutdown waits for async go routines to finish and returns.
+func (p *Peer) Shutdown() {
+	p.mtx.Lock()
+	p.running = false
+	p.mtx.Unlock()
+
+	p.eg.Wait()
```

Comment:
Yup, nice catch! I left it there just in case if we decide to do shutdown with commented code.  I am not sure if we are going to need that, depends on the graceful shutdown. I will switch to wait group for now then and bring back eg if needed!